server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java

```diff
@@ -346,7 +365,114 @@ protected MapperMergeContext createChildContext(MapperMergeContext mapperMergeCo
+        private final Query childFilter;
+
+        private LeafStoredFieldLoader leafStoredFieldLoader;
+        private SourceLoader.Leaf leafSourceLoader;
+        private final List<Integer> children = new ArrayList<>();
+
+        private NestedSyntheticFieldLoader(
+            org.elasticsearch.index.fieldvisitor.StoredFieldLoader storedFieldLoader,
+            SourceLoader sourceLoader,
+            Supplier<BitSetProducer> parentBitSetProducer,
+            Query childFilter
+        ) {
+            this.storedFieldLoader = storedFieldLoader;
+            this.sourceLoader = sourceLoader;
+            this.parentBitSetProducer = parentBitSetProducer;
+            this.childFilter = childFilter;
+        }
+
+        @Override
+        public Stream<Map.Entry<String, StoredFieldLoader>> storedFieldLoaders() {
+            return Stream.of();
+        }
+
+        @Override
+        public DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf) throws IOException {
+            this.children.clear();
+            this.leafStoredFieldLoader = storedFieldLoader.getLoader(leafReader.getContext(), null);
+            this.leafSourceLoader = sourceLoader.leaf(leafReader, null);
+
+            IndexSearcher searcher = new IndexSearcher(leafReader);
+            searcher.setQueryCache(null);
+            var childScorer = searcher.createWeight(childFilter, ScoreMode.COMPLETE_NO_SCORES, 1f).scorer(leafReader.getContext());
+            if (childScorer != null) {
+                var parentDocs = parentBitSetProducer.get().getBitSet(leafReader.getContext());
+                return parentDoc -> {
+                    collectChildren(parentDoc, parentDocs, childScorer.iterator());
+                    return children.size() > 0;
+                };
+            } else {
+                return parentDoc -> false;
+            }
+        }
+
+        private List<Integer> collectChildren(int parentDoc, BitSet parentDocs, DocIdSetIterator childIt) throws IOException {
+            assert parentDocs.get(parentDoc) : "wrong context, doc " + parentDoc + " is not a parent of " + nestedTypePath;
+            final int prevParentDoc = parentDocs.prevSetBit(parentDoc - 1);
+            int childDocId = childIt.docID();
+            if (childDocId <= prevParentDoc) {
+                childDocId = childIt.advance(prevParentDoc + 1);
+            }
+
+            children.clear();
+            for (; childDocId < parentDoc; childDocId = childIt.nextDoc()) {
+                children.add(childDocId);
+            }
+            return children;
+        }
+
+        @Override
+        public boolean hasValue() {
+            return children.size() > 0;
+        }
+
+        @Override
+        public void write(XContentBuilder b) throws IOException {
+            assert (children != null && children.size() > 0);
+            switch (children.size()) {
+                case 1 -> {
```

Comment:
Nit: why use a switch here and not a simple if-else?

Reply:
https://github.com/elastic/elasticsearch/pull/109809/commits/2e3fef7bf86933da05d9c439988c40dc13acab39